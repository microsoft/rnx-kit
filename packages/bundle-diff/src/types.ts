export type BasicSourceMap = {
  sources: Array<string>;
  sourcesContent: Array<string | null>;
};

export type SourceDiff = {
  state: "added" | "changed" | "removed";
  path: string;
  diff: number;
};

export type BundleDiff = {
  files: SourceDiff[];
  total: number;
};
