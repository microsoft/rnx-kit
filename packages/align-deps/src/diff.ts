import { keysOf } from "@rnx-kit/tools-language";
import type { PackageManifest } from "@rnx-kit/tools-node/package";
import type { Changes } from "./types";

export function diff(
  manifest: PackageManifest,
  updatedManifest: PackageManifest
): Changes | undefined {
  const allChanges: Changes = {
    dependencies: [],
    peerDependencies: [],
    devDependencies: [],
  };

  const numChanges = keysOf(allChanges).reduce((count, section) => {
    const changes = allChanges[section];
    const currentDeps = manifest[section] ?? {};
    const updatedDeps = updatedManifest[section] ?? {};

    for (const [dependency, target] of Object.entries(updatedDeps)) {
      const current = currentDeps[dependency];
      if (!current) {
        changes.push({ type: "added", dependency, target });
      } else if (current !== target) {
        changes.push({ type: "changed", dependency, target, current });
      }
    }

    for (const dependency of Object.keys(currentDeps)) {
      if (!updatedDeps[dependency]) {
        changes.push({ type: "removed", dependency });
      }
    }

    return count + changes.length;
  }, 0);

  return numChanges > 0 ? allChanges : undefined;
}

export function stringify(allChanges: Changes, output: string[] = []): string {
  const prefix = "\t  - ";

  for (const [section, changes] of Object.entries(allChanges)) {
    if (changes.length > 0) {
      output.push(`\tIn ${section}:`);
      for (const change of changes) {
        const { type, dependency } = change;
        switch (type) {
          case "added":
            output.push(`${prefix}${dependency} "${change.target}" is missing`);
            break;
          case "changed":
            output.push(
              `${prefix}${dependency} "${change.current}" should be "${change.target}"`
            );
            break;
          case "removed":
            output.push(`${prefix}${dependency} should be removed`);
            break;
        }
      }
    }
  }

  return output.join("\n");
}
