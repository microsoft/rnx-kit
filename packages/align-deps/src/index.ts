export { capabilitiesFor } from "./capabilities";
export { checkPackageManifest } from "./check";
export { cli } from "./cli";
export { updatePackageManifest } from "./manifest";

export type {
  Args,
  CapabilitiesOptions,
  CheckOptions,
  Command,
  DependencyType,
  ExcludedPackage,
  ManifestProfile,
  MetaPackage,
  Package,
  Profile,
  ProfilesInfo,
  ProfileVersion,
  VigilantOptions,
} from "./types";
