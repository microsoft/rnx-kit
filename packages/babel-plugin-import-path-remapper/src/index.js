/* jshint esversion: 8, node: true */
// @ts-check
"use strict";

const { types } = require("@babel/core");
const { declare } = require("@babel/helper-plugin-utils");
const { parseModuleRef } = require("@rnx-kit/tools-node/module");
const { readPackage } = require("@rnx-kit/tools-node/package");

/**
 * @template T
 * @typedef {import("@babel/core").NodePath<T>} NodePath;
 */

/**
 * @typedef {import("@babel/core").types.CallExpression} CallExpression;
 * @typedef {import("@babel/core").types.ExportAllDeclaration} ExportAllDeclaration;
 * @typedef {import("@babel/core").types.ExportNamedDeclaration} ExportNamedDeclaration;
 * @typedef {import("@babel/core").types.ImportDeclaration} ImportDeclaration;
 * @typedef {import("@babel/core").types.Node} Node;
 * @typedef {
 *   | NodePath<ExportAllDeclaration>
 *   | NodePath<ExportNamedDeclaration>
 *   | NodePath<ImportDeclaration>
 * } ImportExportDeclarationNodePath;
 *
 * @typedef {(moduleName: string, path?: string, requester?: string) => string} CustomRemapper;
 */

/**
 * Finds the main source file in the specified package's manifest.
 * @param {string} sourcePath
 * @param {string | undefined} requester
 * @param {CustomRemapper | undefined} customRemap
 * @returns {string | undefined}
 */
function findMainSourceFile(sourcePath, requester, customRemap) {
  const path = require("path");

  const resolveOptions = {
    paths: [requester ? path.dirname(requester) : process.cwd()],
  };
  const manifestPath = require.resolve(
    `${sourcePath}/package.json`,
    resolveOptions
  );

  const { main } = readPackage(manifestPath);
  if (customRemap) {
    return customRemap(
      sourcePath,
      typeof main === "string" ? main : undefined,
      requester
    );
  }

  if (typeof main !== "string") {
    return;
  }

  const remappedPath = `${sourcePath}/${main.replace(
    /^(?:\.\/)?lib\/(.*)\.js/,
    "src/$1.ts"
  )}`;
  try {
    const tsx = `${remappedPath}x`;
    require.resolve(tsx, resolveOptions);
    return tsx;
  } catch (_) {
    return remappedPath;
  }
}

/**
 * Replaces the source string in specified call expression.
 * @param {NodePath<CallExpression>} path
 * @param {string} source
 */
function updateCallWith(path, source) {
  const firstArgument = path.node.arguments[0];
  if (types.isStringLiteral(firstArgument)) {
    firstArgument.value = source;
  }
}

/**
 * Replaces the source string in specified import/export declaration.
 * @param {NodePath<ImportExportDeclarationNodePath>} path
 * @param {string} source
 */
function updateDeclarationWith(path, source) {
  path.node.source.value = source;
}

/**
 * @template T
 * @param {string} sourcePath
 * @param {string | undefined} requester
 * @param {NodePath<T>} path
 * @param {(path: NodePath<T>, source: string) => void} updater
 * @param {CustomRemapper | undefined} customRemap
 */
function update(sourcePath, requester, path, updater, customRemap) {
  const m = parseModuleRef(sourcePath);
  if (!("name" in m)) {
    // This is not a module reference. Ignore unless we have a custom remapper.
    if (customRemap) {
      updater(path, customRemap("", m.path, requester));
    }
    return;
  }

  const { scope, name: moduleName, path: modulePath } = m;
  if (!modulePath) {
    // Remaps @scope/example -> @scope/example/src/index.ts
    try {
      const mainSourceFile = findMainSourceFile(
        sourcePath,
        requester,
        customRemap
      );
      if (mainSourceFile) {
        updater(path, mainSourceFile);
      }
    } catch (_) {
      /* ignore */
    }
    return;
  }

  const name = scope ? `${scope}/${moduleName}` : moduleName;
  if (customRemap) {
    updater(path, customRemap(name, modulePath));
  } else if (modulePath === "lib" || modulePath.startsWith("lib")) {
    // Remaps @scope/example/lib/index.js -> @scope/example/src/index.ts
    updater(path, `${name}/${modulePath.replace("lib", "src")}`);
  }
}

module.exports = declare((api, options) => {
  api.assertVersion(7);

  /**
   * @type {{
   *   test?: (source: string) => boolean;
   *   remap?: CustomRemapper;
   * }}
   */
  const { test, remap } = options;
  if (typeof test !== "function") {
    throw new Error(
      "Expected option `test` to be a function `(source: string) => boolean`"
    );
  }
  if (remap && typeof remap !== "function") {
    throw new Error(
      "Expected option `remap` to be undefined or a function `(moduleName: string, path: string) => string`"
    );
  }

  return {
    name: "import-path-remapper",
    visitor: {
      CallExpression: (path, state) => {
        if (
          !types.isImport(path.node.callee) &&
          (!types.isIdentifier(path.node.callee) ||
            path.node.callee.name !== "require")
        ) {
          return;
        }

        const arg = path.node.arguments[0];
        const sourcePath = arg && types.isStringLiteral(arg) && arg.value;
        if (!sourcePath || !test(sourcePath)) {
          return;
        }

        update(sourcePath, state.filename, path, updateCallWith, remap);
      },

      /** @type {(path: ImportExportDeclarationNodePath, state: { filename?: string; }) => void} */
      "ImportDeclaration|ExportDeclaration": (path, state) => {
        if (!path.node.source) {
          // Ignore non-re-export lines, e.g.: export const example = () => { ... }
          return;
        }

        const sourcePath = path.node.source.value;
        if (!test(sourcePath)) {
          return;
        }

        update(sourcePath, state.filename, path, updateDeclarationWith, remap);
      },
    },
  };
});
