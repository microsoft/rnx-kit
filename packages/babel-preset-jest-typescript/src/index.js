/* jshint esversion: 8, node: true */
// @ts-check
"use strict";

module.exports = () => {
  return {
    presets: [
      ["@babel/preset-env", { targets: { node: "current" } }],
      "@babel/preset-typescript",
    ],
  };
};
