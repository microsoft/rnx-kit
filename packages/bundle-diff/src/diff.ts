import type { BasicSourceMap, SourceDiff } from "./types";

export function isValidSourceMap(
  obj: Partial<BasicSourceMap>
): obj is BasicSourceMap {
  const { sources, sourcesContent } = obj;
  return Array.isArray(sources) && Array.isArray(sourcesContent);
}

export function makeMap({
  sources,
  sourcesContent,
}: BasicSourceMap): Record<string, number> {
  return sources.reduce<Record<string, number>>((map, file, index) => {
    const content = sourcesContent[index];
    map[file] = content?.length ?? NaN;
    return map;
  }, {});
}

export function diff(
  a: Partial<BasicSourceMap>,
  b: Partial<BasicSourceMap>
): SourceDiff[] {
  if (!isValidSourceMap(a) || !isValidSourceMap(b)) {
    return [];
  }

  const aMap = makeMap(a);
  const bMap = makeMap(b);
  return Array.from(
    [Object.keys(aMap), Object.keys(bMap)].reduce((set, sources) => {
      return sources.reduce((set, sourceFile) => {
        set.add(sourceFile);
        return set;
      }, set);
    }, new Set<string>())
  ).reduce<SourceDiff[]>((files, sourceFile) => {
    const aFileSize = aMap[sourceFile];
    const bFileSize = bMap[sourceFile];

    // NaN !== NaN, so we must use `Object.is` to compare them. We still need
    // to use `===` because `Object.is(0, -0)` is `false`.
    if (aFileSize === bFileSize || Object.is(aFileSize, bFileSize)) {
      return files;
    }

    const delta = (bFileSize ?? 0) - (aFileSize ?? 0);
    files.push({
      state: !aFileSize ? "added" : !bFileSize ? "removed" : "changed",
      path: sourceFile,
      diff: delta,
    });

    return files;
  }, []);
}
