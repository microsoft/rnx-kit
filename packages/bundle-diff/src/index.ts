export { cli } from "./cli";
export { diff } from "./diff";
