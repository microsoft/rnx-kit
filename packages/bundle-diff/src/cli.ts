#!/usr/bin/env node
import fs from "fs";
import { diff } from "./diff";

export function withSign(n: number): string {
  if (Number.isNaN(n)) {
    return "unknown";
  }

  if (n === 0) {
    return "Â±0";
  }

  return n > 0 ? `+${n}` : n.toString();
}

export function numDigits(n: number): number {
  if (Number.isNaN(n)) {
    return withSign(n).length;
  }

  if (n === 0) {
    return 1;
  }

  return Math.floor(Math.log10(Math.abs(n)) + 1);
}

export function numDigitsStringLength(content: string | null): number {
  return numDigits(content?.length ?? NaN);
}

export function cli(aSourceMap: string, bSourceMap: string): void {
  const a = JSON.parse(fs.readFileSync(aSourceMap, { encoding: "utf-8" }));
  const b = JSON.parse(fs.readFileSync(bSourceMap, { encoding: "utf-8" }));

  const digits = Math.max(
    ...a.sourcesContent.map(numDigitsStringLength),
    ...b.sourcesContent.map(numDigitsStringLength)
  );

  const output = diff(a, b)
    .sort((a, b) => b.diff - a.diff)
    .map(({ diff, path, state }) => {
      const contentSize = withSign(diff).padStart(digits + 1);
      const sourceState = state.padStart(7);
      return ` ${contentSize}  ${sourceState}  ${path}`;
    })
    .join("\n");

  if (output) {
    console.log(output);
  }
}

if (require.main === module) {
  const { [2]: aSourceMap, [3]: bSourceMap } = process.argv;
  cli(aSourceMap, bSourceMap);
}
