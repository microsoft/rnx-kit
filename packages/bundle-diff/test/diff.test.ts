import { diff, isValidSourceMap, makeMap } from "../src/diff";
import { aSourceMap, bSourceMap } from "./mockSourceMaps";

describe("isValidSourceMap()", () => {
  test("returns true when an object resembles a source map", () => {
    expect(isValidSourceMap({})).toBe(false);
    expect(isValidSourceMap({ sources: [] })).toBe(false);
    expect(isValidSourceMap({ sourcesContent: [] })).toBe(false);

    const badSourceMap = { sources: "", sourcesContent: "" } as any;
    expect(isValidSourceMap(badSourceMap)).toBe(false);

    expect(isValidSourceMap({ sources: [], sourcesContent: [] })).toBe(true);
  });
});

describe("makeMap()", () => {
  test("creates a map of files with content", () => {
    expect(makeMap(aSourceMap)).toMatchSnapshot();
  });
});

describe("diff()", () => {
  test("returns empty array if one of the source maps is invalid", () => {
    expect(diff({}, bSourceMap)).toEqual([]);
    expect(diff(aSourceMap, {})).toEqual([]);
    expect(diff({}, {})).toEqual([]);
  });

  test("returns empty array if there are no diffs", () => {
    expect(diff(aSourceMap, aSourceMap)).toEqual([]);
  });

  test("returns files that differ between two bundles", () => {
    expect(diff(aSourceMap, bSourceMap)).toMatchSnapshot();
  });
});
