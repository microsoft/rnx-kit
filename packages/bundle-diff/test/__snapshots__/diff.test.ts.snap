// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff() returns files that differ between two bundles 1`] = `
Array [
  Object {
    "diff": -127,
    "path": "/~/node_modules/querystring-es3/index.js",
    "state": "removed",
  },
  Object {
    "diff": 1,
    "path": "/~/node_modules/react/index.js",
    "state": "changed",
  },
  Object {
    "diff": -286,
    "path": "/~/node_modules/react-native/Libraries/Components/Picker/PickerAndroid.ios.js",
    "state": "removed",
  },
  Object {
    "diff": -592,
    "path": "/~/node_modules/react-native/Libraries/Components/Sound/SoundManager.js",
    "state": "removed",
  },
  Object {
    "diff": NaN,
    "path": "/~/packages/awesome-app/lib/index.js",
    "state": "added",
  },
  Object {
    "diff": 106,
    "path": "/~/node_modules/@babel/runtime/helpers/arrayWithHoles.js",
    "state": "added",
  },
  Object {
    "diff": 49,
    "path": "/~/node_modules/@babel/runtime/regenerator/index.js",
    "state": "added",
  },
  Object {
    "diff": 96,
    "path": "/~/node_modules/lodash-es/_realNames.js",
    "state": "added",
  },
]
`;

exports[`makeMap() creates a map of files with content 1`] = `
Object {
  "/~/node_modules/querystring-es3/index.js": 127,
  "/~/node_modules/react-native/Libraries/Components/Picker/PickerAndroid.ios.js": 286,
  "/~/node_modules/react-native/Libraries/Components/Sound/SoundManager.js": 592,
  "/~/node_modules/react/index.js": 42,
  "/~/packages/awesome-app/lib/index.js": NaN,
}
`;
